### PR TITLE
[MXNET-978] Support higher order gradient for `log`.

### PR DESCRIPTION
**Pull request by kshitij12345**
## Description ##
With reference to #14613,  #10002 , this PR intends to add support for higher order gradient for `log` { and ideally for `log2, log10` } as well.

Tests are based totally on #14613 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/browse/MXNET-978) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] higher order gradient for a `log`.
- [x] unit test for the same.

### CODE REVIEW

File: src/operator/tensor/elemwise_unary_op_basic.cc [hunk 1]
```diff
@@ -1016,7 +1016,28 @@ The storage type of ``log2`` output is always dense
 .set_attr<nnvm::FGradient>("FGradient", ElemwiseGradUseIn{"_backward_log2"});
 
 MXNET_OPERATOR_REGISTER_BINARY_WITH_SPARSE_CPU_DR(_backward_log,
-                                                  unary_bwd<mshadow_op::log_grad>);
+                                                  unary_bwd<mshadow_op::log_grad>)
```
Comment (kshitij12345) [inline src/operator/tensor/elemwise_unary_op_basic.cc]: I don't know much about this library but,

I believe it would be better to have gradients defined for  existing backward, instead of a differentiable gradient (relying on  autograd machinery) at least on `ops`  where backward is not trivial. It will allow to use existing optimised fused kernels and make sure there is no regression in the backward.

Note: `log` is relatively trivial (single reciprocal). But maybe we may see a performance regression for `sigmoid`, if we do it by relying on autograd machinery instead of the existing `_backward_sigmoid`.

**Comment by kshitij12345** [comment]
I don't know much about this library but,

I believe it would be better to have gradients defined for  existing backward, instead of a differentiable gradient (relying on  autograd machinery) at least on `ops`  where backward is not trivial. It will allow to use existing optimised fused kernels and make sure there is no regression in the backward.

Note: `log` is relatively trivial (single reciprocal). But maybe we may see a performance regression for `sigmoid`, if we do it by relying on autograd machinery instead of the existing `_backward_sigmoid`.

**Comment by kshitij12345** [comment]
Can anyone please point how I can have `1/log(2.0), 1/log(10.0)` multiplied with gradient for `log2, log10`.

**Comment by pinaraws** [comment]
@mxnet-label-bot add[Operator, pr-awaiting-review]

**Comment by apeforest** [comment]
@kshitij12345 Thanks for your contribution. I agree with you it would be better to have gradients defined for existing backward operators. 

I do not fully understand your question of 1/log(2.0), 1/log(10.0) multiplied with gradient for log2, log 10. Could you please elaborate?

**Comment by kshitij12345** [comment]
> I do not fully understand your question of 1/log(2.0), 1/log(10.0) multiplied with gradient for log2, log 10. Could you please elaborate?

Reading it again, I phrased it poorly. Sorry. So actually, plan was to update `gradient` for `log2`, which would be `1/(log(2.0)  * x)`, for which I would have required a `log(2.0)`. So how to get that, is scalar multiplication allowed ? or `ones_like` followed by `fill`.  

**Note**: This is not need for this PR. But curious to know.

Thank You.

**Comment by kshitij12345** [comment]
@larroy @apeforest Have updated as per #14095 .
Please review.

**Comment by kshitij12345** [comment]
Failure looks irrelevant to the PR.

http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/mxnet-validation%2Funix-gpu/detail/PR-14992/7/pipeline#step-767-log-360

File: src/operator/tensor/elemwise_unary_op_basic.cc [hunk 2]
```diff
@@ -1069,13 +1069,72 @@ The storage type of ``log2`` output is always dense
...
+.set_attr<nnvm::FGradient>("FGradient",
+  [](const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& ograds) {
+    // For g(x) -> g = log
+    // g''(x) = -1 * (g'(x) * g'(x))
+    auto gx = nnvm::NodeEntry{n, 0, 0};
```
Comment (larroy) [inline src/operator/tensor/elemwise_unary_op_basic.cc]: You are welcome to simplify calls to NodeEntry as per:

https://github.com/apache/incubator-mxnet/pull/14095

Just call nnvm::NodeEntry{n}

File: src/operator/tensor/elemwise_unary_op_basic.cc [hunk 3]
```diff
@@ -1069,13 +1069,72 @@ The storage type of ``log2`` output is always dense
...
+    auto ggx_mid = MakeNode("elemwise_mul", n->attrs.name + "_backward_mid_grad_grad",
+                            {gx, gx}, nullptr, &n);
+    auto ggx = MakeNode("negative", n->attrs.name + "_backward_grad_grad",
+                        {nnvm::NodeEntry{ggx_mid, 0, 0}}, nullptr, &n);
+
+    std::vector<nnvm::NodeEntry> ret;
+
+    ret.emplace_back(MakeNode("elemwise_mul", n->attrs.name + "_backward_grad_grad",
+                             {ograds[0], gx}, nullptr, &n));
+    ret.emplace_back(MakeNode("elemwise_mul", n->attrs.name + "_backward_grad_grad_inp",
```
Comment (larroy) [inline src/operator/tensor/elemwise_unary_op_basic.cc]: Why are we returning two gradients, isn't it an unary function with just one input?

File: src/operator/tensor/elemwise_unary_op_basic.cc [hunk 4]
```diff
@@ -1069,13 +1069,72 @@ The storage type of ``log2`` output is always dense
...
+    auto g_lx = MakeNode("reciprocal", n->attrs.name + "_backward_log_grad",
+                            {n->inputs[1]}, nullptr, &n);
+    auto ggx_mid = MakeNode("elemwise_mul", n->attrs.name + "_backward_mid_grad_grad",
+                            {gx, nnvm::NodeEntry{g_lx, 0, 0}}, nullptr, &n);
+    auto ggx = MakeNode("negative", n->attrs.name + "_backward_grad_grad",
+                        {nnvm::NodeEntry{ggx_mid, 0, 0}}, nullptr, &n);
+
+    std::vector<nnvm::NodeEntry> ret;
+
+    ret.emplace_back(MakeNode("elemwise_mul", n->attrs.name + "_backward_grad_grad",
```
Comment (larroy) [inline src/operator/tensor/elemwise_unary_op_basic.cc]: same comment as above.

File: src/operator/tensor/elemwise_unary_op_basic.cc (hunk 3 above)
Comment (kshitij12345) [reply src/operator/tensor/elemwise_unary_op_basic.cc]: `log` is unary, but the `_backward_log` isn't. 
https://github.com/apache/incubator-mxnet/blob/016556553ced0c9c9400efec2f08a3c39cbb00ba/src/operator/tensor/elemwise_unary_op_basic.cc#L1071

It takes as input the output gradient and input to log.
https://github.com/apache/incubator-mxnet/blob/016556553ced0c9c9400efec2f08a3c39cbb00ba/src/operator/tensor/elemwise_unary_op_basic.cc#L1045

File: src/operator/tensor/elemwise_unary_op_basic.cc (hunk 2 above)
Comment (kshitij12345) [reply src/operator/tensor/elemwise_unary_op_basic.cc]: Oh. Missed that. Thank You.

**Comment by apeforest** [comment]
> > I do not fully understand your question of 1/log(2.0), 1/log(10.0) multiplied with gradient for log2, log 10. Could you please elaborate?
> 
> Reading it again, I phrased it poorly. Sorry. So actually, plan was to update `gradient` for `log2`, which would be `1/(log(2.0) * x)`, for which I would have required a `log(2.0)`. So how to get that, is scalar multiplication allowed ? or `ones_like` followed by `fill`.
> 
> **Note**: This is not needed for this PR. But curious to know.
> 
> Thank You.

Can you just add an elemwise_mul operator node between them?

File: src/operator/tensor/elemwise_unary_op_basic.cc [hunk 5]
```diff
@@ -1069,13 +1069,73 @@ The storage type of ``log2`` output is always dense
 .set_attr<nnvm::FGradient>("FGradient", ElemwiseGradUseIn{"_backward_log2"});
 
 MXNET_OPERATOR_REGISTER_BINARY_WITH_SPARSE_CPU_DR(_backward_log,
-                                                  unary_bwd<mshadow_op::log_grad>);
+                                                  unary_bwd<mshadow_op::log_grad>)
+.set_attr<nnvm::FGradient>("FGradient",
+  [](const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& ograds) {
+    // For g(x) -> g = log
```
Comment (apeforest) [inline src/operator/tensor/elemwise_unary_op_basic.cc]: nit: It's very nice to see a comment here. The g(x) is actually a function of x. It might be easily confused with the variable gx two lines below. Maybe use `f(x)` in the comment here?

**Comment by apeforest** [comment]
@kshitij12345 The CI failure in unix-GPU was due to a flaky test for TensorRT: https://github.com/apache/incubator-mxnet/issues/14978

The issue has been fixed by https://github.com/apache/incubator-mxnet/pull/15014. Please re-trigger CI again. Thanks!

File: src/operator/tensor/elemwise_unary_op_basic.cc (hunk 5 above)
Comment (kshitij12345) [reply src/operator/tensor/elemwise_unary_op_basic.cc]: Ah sure . That makes sense. Thank You.

**Comment by larroy** [comment]
Lgtm

**Comment by larroy** [comment]
"I believe it would be better to have gradients defined for existing backward, instead of a differentiable gradient (relying on autograd machinery) at least on ops where backward is not trivial. It will allow to use existing optimised fused kernels and make sure there is no regression in the backward."

I think you are right, we have discussed this before. I think we rushed merging this PR. We should do a check for regressions.

File: src/operator/tensor/elemwise_unary_op_basic.cc [hunk 6]
```diff
@@ -1069,13 +1069,73 @@ The storage type of ``log2`` output is always dense
...
 
 MXNET_OPERATOR_REGISTER_BINARY_WITH_SPARSE_CPU_DR(_backward_log10,
-                                                  unary_bwd<mshadow_op::log10_grad>);
+                                                  unary_bwd<mshadow_op::log10_grad>)
+.set_attr<nnvm::FGradient>("FGradient",
+  [](const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& ograds) {
+    // For f(x) -> f = log10
+    // f'(x) = 1 / (log(10) * x)
+    // f''(x) = -1 * (f'(x) * 1/x)
+    auto gx = nnvm::NodeEntry{n, 0, 0};
```
Comment (larroy) [inline src/operator/tensor/elemwise_unary_op_basic.cc]: Why don't we follow the same pattern as in the natural logarithm?

Comment (kshitij12345) [reply src/operator/tensor/elemwise_unary_op_basic.cc]: For natural `log`,
we have with us in `gradient` function, `gx` i.e. `1/x` as well as `x`.
Since, second derivative of log is `-(gx * gx)` = `-1/(x^2)`. We use the pattern.

Considering `log2` (similar case for `log10`)
we have with us, `gx` i.e. `1/(log(2) * x)` as well as `x`.
Since second derivative is `-1/(log(2) * x * x)`
which we get in the code using `negative(gx * reciprocal(x))`, where `gx=1/(log(2) * x`.
Another way to get that will be `negative(gx * gx * log(2.0))`.

Comment (kshitij12345) [reply src/operator/tensor/elemwise_unary_op_basic.cc]: @larroy Thanks for pointing this, going through this again made me realise that there is a problem with the implementation of `log`.

**Comment by kshitij12345** [comment]
@apeforest @larroy 

https://github.com/kshitij12345/incubator-mxnet/blob/7b343d1fcde73b61322985580080333d9eee9e82/src/operator/tensor/elemwise_unary_op_basic.cc#L1077-L1079

We multiply  `gx * gx` where `gx = ograd * f'(x)`, getting `ograd^2 * f'(x)^2`, however we want only `ograd * f'(x)^2` which can be achieved in a similar fashion to the implementation of `_backward_log10/2`.

I have validated the expected results.
```python
from mxnet import nd, autograd
import numpy
import math
grad_grad_op = lambda x: (-1/x**2)

x = nd.random.normal(0,1,(3,3))
x.attach_grad()
with autograd.record():
  y = nd.log(x)
  y_grad = autograd.grad(y, x, head_grads= nd.ones_like(y) * 0.5, create_graph=True, retain_graph=True)[0]
y_grad.backward(nd.ones_like(y_grad) * 0.6)

numpy.testing.assert_allclose(x.grad.asnumpy() , ( grad_grad_op(x) * 0.5 * 0.6).asnumpy(), rtol=1e-7, atol=1e-7)
```
Which fails with current code.
**Should make a new PR, or add commits in this PR itself?**

Have confirmed the behaviour with Pytorch as well.
```python
import torch
import numpy
import math

grad_grad_op = lambda x: (-1/x**2)

x = torch.randn(2,3)
x.requires_grad = True

y = torch.log(x)
y_grad = torch.autograd.grad(y, x, grad_outputs=torch.ones_like(y) * 0.5, create_graph=True, retain_graph=True)[0]
y_grad.backward(torch.ones_like(y_grad) * 0.6)

numpy.testing.assert_allclose(x.grad.detach().numpy() , ( grad_grad_op(x) * 0.5 * 0.6).detach().numpy(), rtol=1e-7, atol=1e-7)
```

**Comment by apeforest** [comment]
Hi @kshitij12345 sorry, we missed that. We should have reviewed it more carefully. Please submit another PR to fix this issue. I will also update https://github.com/apache/incubator-mxnet/pull/14613 accordingly. Thanks!

File: src/operator/tensor/elemwise_unary_op_basic.cc [hunk 7]
```diff
@@ -1069,13 +1069,73 @@ The storage type of ``log2`` output is always dense
...
+                            {n->inputs[1]}, nullptr, &n);
+    auto ggx_mid = MakeNode("elemwise_mul", n->attrs.name + "_backward_mid_grad_grad",
+                            {gx, nnvm::NodeEntry{g_lx}}, nullptr, &n);
+    auto ggx = MakeNode("negative", n->attrs.name + "_backward_grad_grad",
+                        {nnvm::NodeEntry{ggx_mid}}, nullptr, &n);
+
+    std::vector<nnvm::NodeEntry> ret;
+
+    ret.emplace_back(MakeNode("elemwise_mul", n->attrs.name + "_backward_grad_grad",
+                             {ograds[0], gx}, nullptr, &n));
```
Comment (apeforest) [inline src/operator/tensor/elemwise_unary_op_basic.cc]: Shouldn't this be `{ograds[0], g_lx}` instead? Isn't `dL/dy_grad = d^2L/dx^2 * f'(x)`?

Comment (kshitij12345) [reply src/operator/tensor/elemwise_unary_op_basic.cc]: Yes, it should. Thanks. I have updated this change and added relevant test in the new PR #15120 .
Actually I am having trouble exactly at this part as the `grad` value is not being updated. More info in #15120

**Comment by kshitij12345** [comment]
Have created a new PR for the same. #15120 Please review
